modules/apps/27-interchain-accounts/types/codec_test.go

```diff
@@ -41,110 +44,258 @@ func (mockSdkMsg) GetSigners() []sdk.AccAddress {
+					&mockSdkMsg{},
+				}
 			},
 			false,
 		},
 		{
-			"multiple unregistered msg types",
-			[]proto.Message{
-				&mockSdkMsg{},
-				&mockSdkMsg{},
-				&mockSdkMsg{},
+			"failure: multiple unregistered msg types",
+			func() {
+				msgs = []proto.Message{
+					&mockSdkMsg{},
+					&mockSdkMsg{},
+					&mockSdkMsg{},
+				}
 			},
 			false,
 		},
-	}
+		{
+			"failure: nested unregistered msg",
+			func() {
+				mockMsg := &mockSdkMsg{}
+				mockAny, err := codectypes.NewAnyWithValue(mockMsg)
+				suite.Require().NoError(err)
 
-	for _, tc := range testCases {
-		tc := tc
+				msgs = []proto.Message{
+					&govtypes.MsgSubmitProposal{
+						Content:        mockAny,
+						InitialDeposit: sdk.NewCoins(sdk.NewCoin(sdk.DefaultBondDenom, sdkmath.NewInt(5000))),
+						Proposer:       TestOwnerAddress,
+					},
+				}
+			},
+			false,
+		},
+		{
+			"failure: nested array of unregistered msg",
+			func() {
+				mockMsg := &mockSdkMsg{}
+				mockAny, err := codectypes.NewAnyWithValue(mockMsg)
+				suite.Require().NoError(err)
 
-		suite.Run(tc.name, func() {
-			bz, err := types.SerializeCosmosTx(simapp.MakeTestEncodingConfig().Codec, tc.msgs, types.EncodingProtobuf)
-			suite.Require().NoError(err, tc.name)
+				messages := []*codectypes.Any{mockAny, mockAny, mockAny}
 
-			msgs, err := types.DeserializeCosmosTx(simapp.MakeTestEncodingConfig().Codec, bz, types.EncodingProtobuf)
-			if tc.expPass {
-				suite.Require().NoError(err, tc.name)
-			} else {
-				suite.Require().Error(err, tc.name)
-			}
+				propMsg := &govtypesv1.MsgSubmitProposal{
+					Messages:       messages,
+					InitialDeposit: sdk.NewCoins(sdk.NewCoin(sdk.DefaultBondDenom, sdkmath.NewInt(5000))),
+					Proposer:       TestOwnerAddress,
+					Metadata:       "",
+					Title:          "New IBC Gov Proposal",
+					Summary:        "more tokens for all!",
+				}
 
-			for i, msg := range msgs {
-				suite.Require().Equal(tc.msgs[i], msg)
-			}
-		})
+				msgs = []proto.Message{propMsg}
+			},
+			false,
+		},
 	}
 
-	// test serializing non sdk.Msg type
-	bz, err := types.SerializeCosmosTx(simapp.MakeTestEncodingConfig().Codec, []proto.Message{&banktypes.MsgSendResponse{}}, types.EncodingProtobuf)
-	suite.Require().NoError(err)
-	suite.Require().NotEmpty(bz)
+	for _, encoding := range testedEncodings {
+		for _, tc := range testCases {
+			tc := tc
 
-	// test deserializing unknown bytes
-	_, err = types.DeserializeCosmosTx(simapp.MakeTestEncodingConfig().Codec, bz, types.EncodingProtobuf)
-	suite.Require().Error(err) // unregistered type
+			suite.Run(tc.name, func() {
+				tc.malleate()
 
-	// test deserializing unknown bytes
-	msgs, err := types.DeserializeCosmosTx(simapp.MakeTestEncodingConfig().Codec, []byte("invalid"), types.EncodingProtobuf)
-	suite.Require().Error(err)
-	suite.Require().Empty(msgs)
+				bz, err := types.SerializeCosmosTx(simapp.MakeTestEncodingConfig().Codec, msgs, encoding)
+				if encoding == types.EncodingJSON && !tc.expPass {
```

Comment:
Should we add a comment somewhere that for JSON, `tc.expPass` set to false means it'll fail on serialization while for protobuf it will fail on deserialization? Maybe at the start of the test?

Reply:
done